src/Servers/IIS/IIS/src/Core/IO/AsyncIOEngine.cs

```diff
@@ -35,13 +37,96 @@ public ValueTask<int> ReadAsync(Memory<byte> memory)
     }
 
     public ValueTask<int> WriteAsync(ReadOnlySequence<byte> data)
+    {
+        if (SegmentsOverChunksLimit(data))
+        {
+            return WriteDataOverChunksLimit(data);
+        }
+
+        return WriteDataAsync(data);
+    }
+
+    private ValueTask<int> WriteDataAsync(in ReadOnlySequence<byte> data)
     {
         var write = GetWriteOperation();
         write.Initialize(_handler, data);
         Run(write);
         return new ValueTask<int>(write, 0);
     }
 
+    /*
+    * In case the number of chunks is bigger than responseMaxChunks we need to make multiple calls 
+    * to the native api https://docs.microsoft.com/en-us/iis/web-development-reference/native-code-api-reference/ihttpresponse-writeentitychunks-method  
+    * Despite the documentation states that feeding the function with more than 65535 chunks will cause the function to throw an exception, 
+    * it actually seems that 65534 is the maximum number of chunks allowed.
+    * Also, there seems to be a problem when slicing a ReadOnlySequence on segment borders tracked here https://github.com/dotnet/runtime/issues/67607
+    * That's why we only allow 65533 chunks.  
+    */
+    private async ValueTask<int> WriteDataOverChunksLimit(ReadOnlySequence<byte> data)
+    {
+        var hr = 0;
+        ushort segmentsCount = 0;
+        var start = 0;
+        var length = 0;
+        int result;
+
+        foreach (var segment in data)
+        {
+            segmentsCount++;
+            length += segment.Length;
+
+            if (segmentsCount == responseMaxChunks)
+            {
+                result = await WriteDataAsync(data.Slice(start, length));
+
+                SaveResultIfError(result, ref hr);
+
+                segmentsCount = 0;
+                start = length;
+                length = 0;
+            }
+        }
+
+        if (segmentsCount > 0)
+        {
+            result = await WriteDataAsync(data.Slice(start, length));
+
+            SaveResultIfError(result, ref hr);
+        }
+
+        return hr;
+
+        static void SaveResultIfError(int result, ref int hr)
+        {
+            if (result > 0)
+            {
+                hr = result;
+            }
+        }
+    }
+
+    private static bool SegmentsOverChunksLimit(in ReadOnlySequence<byte> data)
+    {
+        if (data.IsSingleSegment)
+        {
+            return false;
+        }
+
+        var count = 0;
```

Comment:
I think this should be changed to check for if (data.Length > (64K * 4K)) instead of looping over segments.

Reply:
Or I guess if `((data.Length >> 12) > 64K)` might be a bit better.

Reply:
I thought about it but I wasn't sure we could assume that each segment is always (or will always be) 4KB, I thought it was an internal implementation detail of the pipelines. Would it be safe to do so?

Reply:
The default segment size is 4K in pipelines and the memory pool size we’re using is hard coded to 4K currently. It’s less of an implementation detail and more of a default. 

cc @halter73

Reply:
I completely agree with this approach and I also agree on the fact that iterating should be more expensive in this case, it's just that I didn't know how safe it was to base my code on an hard coded value which might change without being able to detect such change at compile time. 

I think I can set up some performance tests, maybe comparing the impact of 2k vs 4k for different buffer size that needs to be sent (I expect 2k hitting the async await branch more often, I'm not sure whether or not this is gonna have a bigger impact than to just iterate for example, but this can be another test). 

In case you have some ideas on how what to test let me know.

Reply:
Lets stick with this simple working implementation for now.